frontend/src/component/InitialRedirect.tsx

```diff
@@ -1,37 +1,59 @@
-import { useCallback, useEffect } from 'react';
-import { useNavigate } from 'react-router-dom';
+import { useEffect } from 'react';
+import { type Location, Navigate } from 'react-router-dom';
 import useProjects from 'hooks/api/getters/useProjects/useProjects';
 import { useLastViewedProject } from 'hooks/useLastViewedProject';
 import Loader from './common/Loader/Loader';
-import { getSessionStorageItem, setSessionStorageItem } from 'utils/storage';
+import { useLocalStorageState } from 'hooks/useLocalStorageState';
+import { useAuthUser } from 'hooks/api/getters/useAuth/useAuthUser';
 
-export const InitialRedirect = () => {
-    const { lastViewed } = useLastViewedProject();
-    const { projects, loading } = useProjects();
-    const navigate = useNavigate();
-    const sessionRedirect = getSessionStorageItem('login-redirect');
-
-    // Redirect based on project and last viewed
-    const getRedirect = useCallback(() => {
-        if (projects && lastViewed) {
-            return `/projects/${lastViewed}`;
-        }
+export const useLastViewedPage = (location?: Location) => {
+    const [state, setState] = useLocalStorageState<string>(
+        'lastViewedPage',
+        '/personal',
+        7 * 24 * 60 * 60 * 1000,
+    );
 
-        return '/personal';
-    }, [lastViewed, projects]);
+    useEffect(() => {
+        if (location) {
+            const page = [
+                '/personal',
+                '/projects',
+                '/search',
+                '/playground',
+                '/insights',
+                '/admin',
+            ].find(
+                (page) =>
+                    page === location.pathname ||
+                    location.pathname.startsWith(`/{page}/`),
+            );
+            if (page) {
+                setState(page);
+            }
+        }
+    }, [location]);
 
-    const redirect = () => {
-        navigate(sessionRedirect ?? getRedirect(), { replace: true });
-    };
+    return state;
+};
 
-    useEffect(() => {
-        setSessionStorageItem('login-redirect');
-        redirect();
-    }, [getRedirect]);
+export const InitialRedirect = () => {
+    const { user, loading: isLoadingAuth } = useAuthUser();
+    const { loading: isLoadingProjects } = useProjects();
+    const isLoggedIn = Boolean(user?.id);
+    const lastViewedPage = useLastViewedPage();
+    const { lastViewed: lastViewedProject } = useLastViewedProject();
 
-    if (loading) {
+    if (isLoadingAuth || isLoadingProjects) {
         return <Loader type='fullscreen' />;
     }
 
-    return null;
+    if (!isLoggedIn) {
+        return <Navigate to='/login' replace />;
+    }
+
+    if (lastViewedPage === '/projects' && lastViewedProject) {
+        return <Navigate to={`/projects/${lastViewedProject}`} replace />;
+    }
+
+    return <Navigate to={lastViewedPage} replace />;
```

Comment:
Here's the most important part. We had a lot of issues of things not redirecting properly, and previous code was tangled up